### PR TITLE
Restart workflow after activity failure instead of quarantining

### PR DESCRIPTION
**Pull request by lmossman**
## What
As was discussed in several zoom meetings, the existing quarantining behavior provides us with little benefit, it is very difficult to determine what is quarantined, and fixing quarantined workflows manually is a cumbersome task that usually involves restarting non-quarantined workflows unnecessarily.

This PR implements the behavior that we discussed, where after an activity fails the maximum number of retries, then the temporal workflow for that connection is restarted. This means that connections will be able to recover automatically from system-wide failures after they have been resolved, without requiring manual intervention.

With our attempt retry logic to have a delay between retrying failed attempts, our temporal workflows should be able to handle these retrying workflows without being overloaded with quick retries. 
Therefore, this should bring us into a better state where we no longer have to manually un-stick workflows after fixing an issue - they should be able to recover automatically.

## How
The main change in this PR is in the ConnectionManagerWorkflowImpl class, in the runMandatoryActivityWithOutput method. Here I have removed the quarantining logic and instead added a delay, after which the workflow is restarted. This delay will protect us from situations in the future where an activity is added that does not have retry logic, and fails immediately. The delay means that in situations like that, the temporal workflow will not be immediately spammed with restarts, as there will be a 10-minute delay between each failure of the activity before the workflow is retried.

The other changes are explained in the comments on the PR.

## Recommended reading order
1. ConnectionManagerWorkflowImpl.java
2. TemporalUtils.java
3. The rest is tests and updates to env configs.

### CODE REVIEW

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalUtils.java
Comment (lmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalUtils.java:70]: This means that the maximum interval between attempt retries will default to 10 minutes, which prevents the exponential backoff from causing attempts to have hours of delay between retries which would slow down automatic system recovery.

These new settings mean that activities will be retried with the following logic:
```
Attempt 1 fails
(wait 30s)
Attempt 2 fails
(wait 1m)
Attempt 3 fails
(wait 2m)
Attempt 4 fails
(wait 4m)
Attempt 5 fails
(wait 8m)
Attempt 6 fails
(wait 10m)
Attempt 7 fails
(wait 10m)
Attempt 8 fails
(wait 10m)
Attempt 9 fails
(wait 10m)
Attempt 10 fails
(execute workflow restart logic)
```
Adding up all those above times gives us a total timespan of 55.5 minutes over which an activity will be retried 10 times, before the workflow is restarted.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java
Comment (lmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:176]: Had to move this down below the statements that transfer values from the `connectionUpdaterInput` into the `workflowState`, so that if the `ensureCleanJobState` activity fails, the resetConnection flag is still set as desired on the workflowState, as that is used in the workflow retrying logic below.

File: airbyte-workers/src/test/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowTest.java
Comment (lmossman) [inline airbyte-workers/src/test/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowTest.java:1338]: I removed this test because I was having trouble getting the `testEnv.sleep()` to work correctly with the sleeping workflow time and with my new reset logic delays. I figured that removing this test is okay because it was just testing the behavior of an activity that fails _after_ the SyncWorkflow, and the above test that I updated tests the behavior of an activity that fails _before_ the SyncWorkflow, which should both be the same.

File: airbyte-workers/src/test/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowTest.java
Comment (lmossman) [inline airbyte-workers/src/test/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowTest.java:1456]: Similar motivation around removing this test - the behavior is already being tested by the `testWorkflowRestartedAfterFailedActivity()` test above

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java
Comment (lmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:108]: The CheckConnection activity that is launched before a sync workflow was currently using a different set of activity options than the other activities launched by the connection manager workflow. Specifically, it was configured to not have any retry logic, since it was reusing the options from the "check connection" that is launched by the UI.

This change just makes the options used for this activity consistent with the other activities in this workflow, so that they all have the same retry behavior.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java
Comment (lmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:490]: The changes to this method are the main changes of this PR

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java
Comment (benmoriceau) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:108]: FYI @evantahler

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java
Comment (benmoriceau) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:506]: Do we want to keep this? given the delay I think it will be complicated to reach this state

Comment (lmossman) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:506]: I don't feel too strongly about this - I thought it may be useful in debugging situations if someone wants to force the activity to be retried instead of waiting for the workflow to restart.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalUtils.java
Comment (davinchia) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalUtils.java:70]: Where is the logic that increases the wait time between failed activity attempts?

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java
Comment (davinchia) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:506]: why is it complicated to reach this state?

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalUtils.java
Comment (davinchia) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalUtils.java:70]: I'm thinking we should lower the retries down to 5, which gives us about 15 mins until a workflow is restarted. 

55 mins seems like a long one restart cycle loop.

Comment (lmossman) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalUtils.java:70]: I think 5 retries makes sense - I will make that change

Comment (lmossman) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalUtils.java:70]: I have pushed a change to update the default value for this `ACTIVITY_MAX_ATTEMPT` env var from 10 to 5

**Comment by lmossman** [comment]
@benmoriceau @davinchia, or any others I tagged on this PR - based on our zoom call it sounds like we have alignment on this approach for some immediate improvements. 

If so, I can merge this in once one of you approve it!

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java
Comment (benmoriceau) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:506]: It was about the long retries that were taking potentially more than 5 hours, The extra 10 minutes sound not needed.

**Review by benmoriceau — APPROVED** [review]

**Comment by benmoriceau** [comment]
> @benmoriceau @davinchia, or any others I tagged on this PR - based on our zoom call it sounds like we have alignment on this approach for some immediate improvements.
> 
> If so, I can merge this in once one of you approve it!

Yeah LGTM

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java
Comment (lmossman) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:506]: Yeah I think that is not as applicable now, because we have adjusted the retries to take ~15 mins max now. This 10 minute wait is more of a failsafe in case someone adds an activity in the future that does not have any retry configuration - in that case, this wait will help protect us from spamming temporal with restarts of the connection very quickly

**Pull request merged**
